OSDOCS#9001: Adding vSphere credentials to Agent (#70551)
openshift/openshift-docs: modules/installation-configuration-parameters.adoc

Author: I used the following logic to selectively include existing vSphere config parameters in the Agent docs:

- The entire vSphere section, which previously was included only in the vSphere configuration page, has been modified to be included in both the vSphere and Agent configuration pages ([here](https://github.com/openshift/openshift-docs/pull/70551/files#diff-14ebe576efb84b2f58b32a0de2996fced237204ae9b091e3cafbd40b65c1ef31R2423) and [here](https://github.com/openshift/openshift-docs/pull/70551/files#diff-14ebe576efb84b2f58b32a0de2996fced237204ae9b091e3cafbd40b65c1ef31R2648)) 
- Within that section of vSphere content, I used vSphere-only conditionals to exclude parameters that only belong in the vSphere page and not the Agent page. ([Here's an example](https://github.com/openshift/openshift-docs/pull/70551/files#diff-14ebe576efb84b2f58b32a0de2996fced237204ae9b091e3cafbd40b65c1ef31R2484-R2498) of content that remains in the vSphere page but does not appear in the Agent page)

--
endif::ibm-cloud[]

> ifdef::agent,vsphere[]

Reviewer: I'm not sure I understand why you opened this ifdef here and then also on line 2616. I think my question is do you need line 2616? Just checking/learning: is it because you close this opening one down on line 2958?

Author: So technically this ifdef and the one on line 2616 are different because they check for slightly different conditions. This ifdef, which is closed on line 2958, basically says "Rule #1: all of the content within these lines can be included if `agent OR vsphere` is true, unless there's an exception".

The ifdef on line 2616 (which closes on line 2622) is basically treated like an exception to that `agent OR vsphere` condition, because it `ONLY` includes content for agent. It's basically saying "here's an exception to rule #1, the content from lines 2616 to 2622 can be included `ONLY` if `agent` is true, but can't be included in vsphere". Let me know if that makes sense, or if you'd like me to explain further.

TLDR the ifdef on line 2606 makes a general rule and the ifdefs like the ones on lines 2616 and 2623 make exceptions to that rule in order to be more specific

Reviewer: I'm picking it up now. There just may be hope for me yet 😄 . Merging

    failureDomains:
|Establishes the relationships between a region and zone. You define a failure domain by using vCenter objects, such as a `datastore` object. A failure domain defines the vCenter location for {product-title} cluster nodes.
|String
endif::vsphere[]
ifdef::agent[]

|platform:
  vsphere:
    failureDomains:
|Establishes the relationships between a region and zone. You define a failure domain by using vCenter objects, such as a `datastore` object. A failure domain defines the vCenter location for {product-title} cluster nodes.

> |An array of vCenter configuration objects.

Reviewer 2: I think vCenter is not correct here.

```suggestion
|An array of failure domain configuration objects.
```

|platform:
  vsphere:
    failureDomains:
      name:

> |The name of the vCenter server.

Reviewer 3: I think it should be 'failure domain name' instead of 'vCenter server'.

====
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        resourcePool:
|Optional. The absolute path of an existing resource pool where the installation program creates the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.
If you do not specify a value, resources are installed in the root of the cluster, for example `/example_datacenter/host/example_cluster/Resources`.
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        folder:
|Optional. The absolute path of an existing folder where the installation program creates the virtual machines, for example, `/<datacenter_name>/vm/<folder_name>/<subfolder_name>`.
If you do not provide this value, the installation program creates a top-level folder in the datacenter virtual machine folder that is named with the infrastructure ID.
If you are providing the infrastructure for the cluster and you do not want to use the default `StorageClass` object, named `thin`, you can omit the `folder` parameter from the `install-config.yaml` file.
|String

> 

Reviewer 4: see https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installing-vsphere-installer-provisioned-customizations.html#configuring-vsphere-regions-zones_installing-vsphere-installer-provisioned-customizations for the schema to use for these fields.

The list of datacenters must match the list of datacenters specified in the `vcenters` field.
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        datastore:
|The path to the vSphere datastore that holds virtual machine files, templates, and ISO images.
[IMPORTANT]
====
You can specify the path of any datastore that exists in a datastore cluster.
By default, Storage vMotion is automatically enabled for a datastore cluster.
Red{nbsp}Hat does not support Storage vMotion, so you must disable Storage vMotion to avoid data loss issues for your {product-title} cluster.

If you must specify VMs across multiple datastores, use a `datastore` object to specify a failure domain in your cluster's `install-config.yaml` configuration file.
For more information, see "VMware vSphere region and zone enablement".
====
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        resourcePool:
|Optional. The absolute path of an existing resource pool where the installation program creates the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.

> If you do not specify a value, resources are installed in the root of the cluster, for example `/example_datacenter/host/example_cluster/Resources`.

Reviewer 3: For agent-based installations, users must specify the resource pool where the VMs are created because the installer does not create the VMs in this scenario.

Author: If I'm interpreting this correctly, I'll get rid of the last line about "not specifying a value". I'm assuming it's still technically an Optional parameter for ABI, but let me know if that's not the case.

|String

|platform:
  vsphere:
    failureDomains:
      topology:
        folder:
|Optional. The absolute path of an existing folder where the installation program creates the virtual machines, for example, `/<datacenter_name>/vm/<folder_name>/<subfolder_name>`.
If you do not provide this value, the installation program creates a top-level folder in the datacenter virtual machine folder that is named with the infrastructure ID.

> If you are providing the infrastructure for the cluster and you do not want to use the default `StorageClass` object, named `thin`, you can omit the `folder` parameter from the `install-config.yaml` file.

Reviewer 3: For agent-based installations, users must specify the folder where the VMs are created because the installer does not create the VMs in this scenario.

Author: If I'm interpreting this right, I'll get rid of line 2726 about "not providing the value". I'm assuming the line below it can stay and that this parameter is still Optional for ABI, but let me know if that's not the case.

      topology:
        computeCluster:
|The path to the vSphere compute cluster.
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        datacenter:
|Lists and defines the datacenters where {product-title} virtual machines (VMs) operate.
The list of datacenters must match the list of datacenters specified in the `vcenters` field.
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        datastore:
|The path to the vSphere datastore that holds virtual machine files, templates, and ISO images.
[IMPORTANT]
====
You can specify the path of any datastore that exists in a datastore cluster.
By default, Storage vMotion is automatically enabled for a datastore cluster.
Red{nbsp}Hat does not support Storage vMotion, so you must disable Storage vMotion to avoid data loss issues for your {product-title} cluster.

If you must specify VMs across multiple datastores, use a `datastore` object to specify a failure domain in your cluster's `install-config.yaml` configuration file.
For more information, see "VMware vSphere region and zone enablement".
====
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        resourcePool:

> |Optional. The absolute path of an existing resource pool where the installation program creates the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.

Reviewer 3: In case of ABI, the user creates the machines, not the installer program.
```suggestion
|Optional. The absolute path of an existing resource pool where the user has created the virtual machines, for example, `/<datacenter_name>/host/<cluster_name>/Resources/<resource_pool_name>/<optional_nested_resource_pool_name>`.
```

// Commenting out the line below because it doesn't apply to Agent, but it may be needed when this content is brought into the regular vSphere docs.
// If you do not specify a value, resources are installed in the root of the cluster, for example `/example_datacenter/host/example_cluster/Resources`.
|String

|platform:
  vsphere:
    failureDomains:
      topology:
        folder:
|Optional. The absolute path of an existing folder where the installation program creates the virtual machines, for example, `/<datacenter_name>/vm/<folder_name>/<subfolder_name>`.
// Commenting out the line below because it doesn't apply to Agent, but it may be needed when this content is brought into the regular vSphere docs.
// If you do not provide this value, the installation program creates a top-level folder in the datacenter virtual machine folder that is named with the infrastructure ID.

> If you are providing the infrastructure for the cluster and you do not want to use the default `StorageClass` object, named `thin`, you can omit the `folder` parameter from the `install-config.yaml` file.

Reviewer 3: In case of ABI, the user creates the machines, not the installer program. Also, I am unable to create an ISO without providing the folder, so it is a required field. Since it is a required field for ABI, we also need to modify this line "you can omit the `folder` parameter from the `install-config.yaml` file."
cc: @rwsu

Author: Sounds good, I modified the first line to say "folder where the user has created the virtual machines". 

Since the last sentence is based on omitting a required parameter, which we can't do, should I get rid of the sentence entirely, or is there a different way to perform the functionality described there?

Author: Per conversation on Slack, I'll remove that last sentence entirely and double check with Richard once it's removed.

Reviewer 2: Manoj is correct. folder is not optional and is a required parameter. I think line 2729 should be commented out.

      datacenters:
|Lists and defines the datacenters where {product-title} virtual machines (VMs) operate. The list of datacenters must match the list of datacenters specified in the `failureDomains` field.
|String

|platform:
  vsphere:
    vcenters:

>       password:

Reviewer 4: each vCenter is a separate entry in the vcenters slice. this allows multiple vCenters to be defined potentially(though its not supported at the moment).
```suggestion
      - password: <pw>
        port: 443
        server: <vcenter fqdn>
```

Author: That makes total sense. Each param is depicted this way because the install writers changed the old config format in the reference tables (e.g. `platform.vsphere.vcenters.password`) to this newer one that saves horizontal space. We're relying on the description/values column to convey things like what's optional and what needs to go together as a set. 

For example, for the compute section of the install-config, instead of documenting `architecture`, `name`, `platform`, etc like this, we rely on the value field of `compute.architecture` to describe the structure: "[Array of MachinePool objects](https://docs.openshift.com/container-platform/4.14/installing/installing_vsphere/installation-config-parameters-vsphere.html#installation-configuration-parameters-optional_installation-config-parameters-vsphere:~:text=the%20compute%20nodes.-,Array%20of%20MachinePool%20objects.,-compute%3A%0A%20%20architecture%3A)"

    vcenters:
|Lists any fully-qualified hostname or IP address of a vCenter server.
|String
endif::vsphere[]
ifdef::agent[]

|platform:
  vsphere:
    vcenters:

> |Configures the connection details for services to communicate with vCenter.

Reviewer 5: Maybe "Configures the connection details so that services can communicate with vCenter"?

> |Configures the connection details for services to communicate with vCenter.

Author: I think either of your suggestions work, although I'm not familiar enough with the content to know if the first one is still totally accurate. I think I prefer your second suggestion and will go with that one.
